pkg/client/limiter.go

```diff
@@ -0,0 +1,86 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"context"
+	"net/url"
+
+	"k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+)
+
+// Without a limit, the adapter could flood the Prometheus API with many
+// requests when there are many pods running in the cluster because a query for
+// getting all pod metrics would translate into (2 x pod number) queries to the
+// Prometheus API.
+// In the worst case, the Prometheus pods can hit their SOMAXCONN limit leading
+// to timed-out requests from other clients. In particular it has been reported
+// that it could fail the Kubelet liveness probes and lead to Prometheus pod
+// restarts.
+// The number has been chosen from empirical data.
+const maxConcurrentRequests = 100
```

Comment:
no power of 2, I'm disappointed :)

At worst a Prometheus pod will serve both prom-adapter (if the other is down or LB missed it) or even 3 or 4 in some weird rolling scenarios which results in 4XX connections < --web.max-connections=512, which leaves room in that queue and the other ones for the other clients.

Reply:
Yes, had some similar cases as well while testing, didn't dig into how LB is set.
I think 100 is a good value.